### PR TITLE
Update interface of `revertibleRandom`

### PR DESCRIPTION
**Pull request by darkdrag00nv2**
Work towards #2712

## Description

1. Introduced a new type `FixedSizeUnsignedInteger`
2. Update the interface of the `revertibleRandom` function to `fun revertibleRandom<T: FixedSizeUnsignedInteger>([modulo: T]): T`

Note that updating the interface is a **breaking change** as the existing usage of the function such as below fails.

```
let rand1 = revertibleRandom()  // error: cannot infer type parameter: `T`
```

#### Noteworthy Points
1. Updating the interface is a breaking change as pointed out in the description and also the issue https://github.com/onflow/cadence/issues/2712#issuecomment-1785810009
2. Since we don't support type inference of function arguments, we will run into the same issues as the `InclusiveRange` constructor function: https://github.com/onflow/cadence/issues/2886

Basically:

```
let r = revertibleRandom<Int8>(0) // checker error, type mismatch - expected: Int8, Found - Int
let r = revertibleRandom<Int8>(Int8(0)) // success
```
3. The modulo operation hasn't been implemented in this PR. It'll be done in a future PR based on the discussion on the issue.
______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Added appropriate labels

### CODE REVIEW

**Comment by darkdrag00nv2** [comment]
Opened up for some early feedback. I haven't yet done the modulo part.

A couple more things to point out:
1. Updating the interface is a breaking change as pointed out in the description and also the issue https://github.com/onflow/cadence/issues/2712#issuecomment-1785810009
2. Since we don't support type inference of function arguments, we will run into the same issues as the `InclusiveRange` constructor function: https://github.com/onflow/cadence/issues/2886

Basically:

```
let r = revertibleRandom<Int8>(0) // checker error, type mismatch
let r = revertibleRandom<Int8>(Int8(0)) // success
```

File: runtime/stdlib/random.go [hunk 1]
```diff
@@ -34,37 +35,164 @@ NOTE: The use of this function is unsafe if not used correctly.
...
+					func() uint64 {
+						return binary.LittleEndian.Uint64(getRandomBytes(generator, 8))
+					},
+				)
+			case interpreter.PrimitiveStaticTypeWord128:
+				return interpreter.NewWord128ValueFromBigInt(
+					inter,
+					func() *big.Int {
+						buffer := getRandomBytes(generator, 16)
+						return interpreter.LittleEndianBytesToUnsignedBigInt(buffer)
```
Comment (SupunS) [inline runtime/stdlib/random.go:180]: Are the bytes returned from `getRandomBytes`/ `generator.ReadRandom()` always little-endian (or is it a safe assumption)?
The API doesn't specify this. cc: @tarakby

**Review by SupunS — COMMENTED** [review]
Changes look good so far!

Given we are introducing a new type, we should probably add some more high-level Cadence tests (you've already added low-level Go tests, which is great!), for usages of `FixedSizeUnsignedIntegerType` in various places, such as:
- As a type annotation for variable
- Constructing a value (valid/invalid)
- `main` function arguments/return type (for import/export)
- Storing to the storage and reading from it.
- Casting, sub-type assignment, etc.

**Comment by SupunS** [comment]
Given this is a breaking change, can you please rebase this to the `feature/stable-cadence` branch?

File: runtime/stdlib/random.go (hunk 1 above)
Comment (tarakby) [reply runtime/stdlib/random.go:180]: `ReadRandom` only generates bytes, it doesn't generate numbers. This means that there is not endianness notion at the level of `ReadRandom` and its output (that's why the doc doesn't talk about it).

Endianness only makes sense when we want to convert the bytes into a number (at `revertibleRandom` implementation level, since it currently returns a 64-bits number). For that purpose, I [chose little endianness](https://github.com/onflow/cadence/blob/25e5405ecf9a1a2c580155012975b1764281fc24/runtime/stdlib/random.go#L65) but we could also choose big. That choice changes the output numbers, but does not impact the random number distribution.

Comment (SupunS) [reply runtime/stdlib/random.go:180]: oh I see. Thanks for the explanation!

@darkdrag00nv2 Maybe we could avoid the conversion here (in `LittleEndianBytesToUnsignedBigInt`) and directly use `BigEndianBytesToUnsignedBigInt`?

Comment (SupunS) [reply runtime/stdlib/random.go:180]: Might also make sense to consider big-endian for all the byte-to-number conversion in that case, for consistency?

File: runtime/sema/type_test.go [hunk 2]
```diff
@@ -892,9 +892,18 @@ func TestCommonSuperType(t *testing.T) {
...
 				},
 				expectedSuperType: IntegerType,
 			},
+			{
+				name: "fixed size unsigned integers",
+				types: []Type{
+					UInt8Type,
+					UInt128Type,
+				},
+				expectedSuperType: FixedSizeUnsignedIntegerType,
```
Comment (turbolent) [inline runtime/sema/type_test.go:905]: 👌

File: runtime/stdlib/random.go (hunk 1 above)
Comment (turbolent) [reply runtime/stdlib/random.go:180]: Agreed! The `binary.(Little|Big)Endian.UintN` functions are useful to convert the bytes into an integer, but it doesn't matter which we use. Maybe consistently use big, and remove `LittleEndianBytesToUnsignedBigInt`

**Review by turbolent — COMMENTED** [review]
Great work! 

Looks good, only thing left is the removal of the unnecessary endianness conversion

File: runtime/stdlib/random.go (hunk 1 above)
Comment (darkdrag00nv2) [reply runtime/stdlib/random.go]: Done

**Comment by tarakby** [comment]
@turbolent, @SupunS, @darkdrag00nv2  do you think we should add the optional parameter `modulo` as part of this PR?
The optional parameter could be added but the implementation of `revertibleRandom` would ignore it for now. This would allow me to open a follow-up PR and use the parameter to generate an output less than `modulo`. 

Another option is to leave this PR as it is, open a new PR where the optional parameter is added, and I'll continue to commit on that new PR. 
Feel free to choose either way, but it would be good to continue the work on this topic now that it's started.

**Comment by turbolent** [comment]
@tarakby either way works. The PR has the `modulo` parameter in the definition, but the implementation still ignores it: https://github.com/onflow/cadence/pull/2943/files#diff-a1bff80973b499df2d90ec34d5ef987dae71f400bc6a8fab5c6fc5856661fcf7R94. We can merge it as-is and address it in a follow-up PR

**Comment by turbolent** [comment]
There's a data race in the tests: https://github.com/onflow/cadence/actions/runs/7121193421/job/19389931602?pr=2943, looking into it

**Comment by tarakby** [comment]
@turbolent, Nice! I didn't notice the parameter is already there. I'll create an issue for myself to work on the remaining part.

File: runtime/stdlib/random.go [hunk 3]
```diff
@@ -34,37 +35,164 @@ NOTE: The use of this function is unsafe if not used correctly.
...
-				},
-			)
+			inter := invocation.Interpreter
+
+			// TODO: Check if invocation has an argument and implement modulo operation.
+
+			returnIntegerType := invocation.TypeParameterTypes.Oldest().Value
+			returnIntegerStaticType := interpreter.ConvertSemaToStaticType(inter, returnIntegerType)
+
+			switch returnIntegerStaticType {
```
Comment (turbolent) [inline runtime/stdlib/random.go:99]: Avoid the unnecessary conversion here and simply switch on the sema type instead

File: runtime/tests/interpreter/equality_test.go [hunk 4]
```diff
@@ -225,6 +225,70 @@ func TestInterpretEqualityOnNumericSuperTypes(t *testing.T) {
...
+			interpreter.PrimitiveStaticTypeWord64,
+			interpreter.PrimitiveStaticTypeWord128,
+			interpreter.PrimitiveStaticTypeWord256,
+		}
+
+		for _, subtype := range subtypes {
+			rhsType := interpreter.PrimitiveStaticTypeUInt8
+			if subtype == rhsType {
+				rhsType = interpreter.PrimitiveStaticTypeWord128
+			}
```
[the remark this answers is not in the document]
Comment (darkdrag00nv2) [reply runtime/tests/interpreter/equality_test.go:250]: Not strictly required but just to keep the static type of `x` and `y` different. We are doing the same in the `Integer subtypes` test just above.

```
  let x: FixedSizeUnsignedInteger = 5 as %s
  let y: FixedSizeUnsignedInteger = 2 as %s
```

File: runtime/stdlib/random.go (hunk 3 above)
Comment (darkdrag00nv2) [reply runtime/stdlib/random.go]: Done. https://github.com/onflow/cadence/pull/2943/commits/1def84b8a5b61e20a54be4094264da9792bd26e8

**Comment by darkdrag00nv2** [comment]
I believe this should labelled with `Language Breaking Change` as the interface is not backwards compatible.

**Comment by SupunS** [comment]
Not sure why this got closed. This should now be opened against the master (https://github.com/onflow/cadence/pull/2971)

**Comment by dsainati1** [comment]
Same, it seems as though it was closed automatically when `feature/stable-cadence` was merged into `master`. I can't re-open it for some reason though.
